src/FSharp.Core/array.fs

```diff
@@ -2217,10 +2241,28 @@ module Array =
             )
             |> ignore
 
+            trueLength, isTrue
+
+        [<CompiledName("Filter")>]
+        let filter predicate (array: 'T[]) =
+            let trueLength, isTrue = countAndCollectTrueItems predicate array
+            let res = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked trueLength
+            let mutable resIdx = 0
+
+            for i = 0 to isTrue.Length - 1 do
```

Comment:
Wondering if this could be parallelized more. Like, count true items in segments (1 per CPU core), do a prefix sum and write values in parallel to the resulting array. That could be faster for cases when there is a large number of true items. 

Same for `partition` - might have even better results there since we always have to write all items.